src/main/java/io/temporal/internal/statemachines/MutableSideEffectStateMachine.java

```diff
@@ -0,0 +1,298 @@
+              .add(
+                  State.RESULT_NOTIFIED_REPLAYING,
+                  ExplicitEvent.NON_MATCHING_EVENT,
+                  State.SKIPPED_NOTIFIED,
+                  InvocationStateMachine::cancelCommandNotifyCachedResult)
+              .add(
+                  State.RESULT_NOTIFIED_REPLAYING,
+                  EventType.EVENT_TYPE_MARKER_RECORDED,
+                  new State[] {State.MARKER_COMMAND_RECORDED, State.SKIPPED_NOTIFIED},
+                  InvocationStateMachine::notifyFromEvent);
+
+  /** Represents a single invocation of mutableSideEffect. */
+  private class InvocationStateMachine
+      extends EntityStateMachineInitialCommand<State, ExplicitEvent, InvocationStateMachine> {
+
+    private final Functions.Proc1<Optional<Payloads>> resultCallback;
+    private final Functions.Func1<Optional<Payloads>, Optional<Payloads>> func;
+
+    InvocationStateMachine(
+        Functions.Func1<Optional<Payloads>, Optional<Payloads>> func,
+        Functions.Proc1<Optional<Payloads>> callback,
+        Functions.Proc1<StateMachine> stateMachineSink) {
+      super(
+          STATE_MACHINE_DEFINITION,
+          MutableSideEffectStateMachine.this.commandSink,
+          stateMachineSink);
+      this.func = Objects.requireNonNull(func);
+      this.resultCallback = Objects.requireNonNull(callback);
+    }
+
+    State getExecutionState() {
+      return replaying.apply() ? State.REPLAYING : State.EXECUTING;
+    }
+
+    @Override
+    public WorkflowStateMachines.HandleEventStatus handleEvent(
+        HistoryEvent event, boolean hasNextEvent) {
+      if (event.getEventType() != EventType.EVENT_TYPE_MARKER_RECORDED
+          || !event
+              .getMarkerRecordedEventAttributes()
+              .getMarkerName()
+              .equals(MUTABLE_SIDE_EFFECT_MARKER_NAME)) {
+        explicitEvent(ExplicitEvent.NON_MATCHING_EVENT);
+        return WorkflowStateMachines.HandleEventStatus.NOT_MATCHING_EVENT;
+      }
+      Map<String, Payloads> detailsMap = event.getMarkerRecordedEventAttributes().getDetailsMap();
+      Optional<Payloads> idPayloads = Optional.ofNullable(detailsMap.get(MARKER_ID_KEY));
+      String expectedId = dataConverter.fromPayloads(0, idPayloads, String.class, String.class);
+      if (Strings.isNullOrEmpty(expectedId)) {
+        throw new IllegalStateException(
+            "Marker details map missing required key: " + MARKER_ID_KEY);
+      }
+      if (!id.equals(expectedId)) {
+        //        explicitEvent(ExplicitEvent.NON_MATCHING_EVENT);
+        return WorkflowStateMachines.HandleEventStatus.NOT_MATCHING_EVENT;
+      }
+      super.handleEvent(event, hasNextEvent);
+      return WorkflowStateMachines.HandleEventStatus.OK;
+    }
+
+    State createMarker() {
+      State toState;
+      RecordMarkerCommandAttributes markerAttributes;
+      Optional<Payloads> updated;
+      updated = func.apply(result);
+      if (!updated.isPresent()) {
```

Comment:
my own learning: i assume that isPresent returns false is the value is identical to the previously stored version, right?

Reply:
right